### DISCUSSION TITLE
Can't remove the watermark in the document, I hope you can help me

### DISCUSSION THREAD

**Discussion by rexyan** [pymupdf/PyMuPDF#3627] [Looking for help] [answered]
I tried to find a solution in the previous discussion. At present, I tried to look for /Watermark tag in xref_object(xref), but did not find this tag.

I tried the add_redact_annot and add_rect_annot methods, but the following text was blocked and did not display correctly.

I have seen in other discussion that there are various ways to add pdf watermarks, may I ask how to solve this format?

Thank you and look forward to your reply

[测试4.pdf](https://github.com/user-attachments/files/15986864/4.pdf)

**Comment by JorjMcKie** [comment, marked as answer]
Bad luck!
In your case, the PDF creator did not create a special object for watermarking, but simply wrote gray regular text across the other text on the page.
There is no safe way to remove it.

**Reply by rexyan** [reply]
Is there any other way to remove it? This is very important to me. Thank you

**Reply by JorjMcKie** [reply]
Well you can look at the end of the `/Contents` stream. Those "watermark" characters are written there.
Then develop a script that replaces `b"BT .... ET"` by `b"BT ET"` in all cases that you do not like to see.
The two byte strings BT / ET mean begin text / end text.
What you would be doing is replacing the operation with a no-op.
